[otbn,mai] Document and fix MAI (#30505)
lowRISC/opentitan: hw/ip/otbn/doc/theory_of_operation.md

Therefore, any OTBN program must check for `KMAC_STATUS.READY = 1` before starting a session.
If a further secure wipe occurs before the termination completes, the `KMAC_DATA_S0` / `KMAC_DATA_S1` WSRs are wiped again while the termination process continues unaffected.

### Masking accelerator interface

OTBN features a masking accelerator interface (MAI) which is based on CSRs/WSRs and connects to a hardware based masking accelerator (MA).

> This masking accelerator is formally verified to be save against first-order power side-channel attacks under the transition- and glitch-extended probing model.

Reviewer: ```suggestion
This masking accelerator is formally verified to be safe against first-order power side-channel attacks under the transition- and glitch-extended probing model.
```

> This masking accelerator is formally verified to be save against first-order power side-channel attacks under the transition- and glitch-extended probing model.

Reviewer 2: Before talking about formal verification, you need to explain what the purpose of the accelerator is. For example:
```
OTBN features a masking accelerator interface (MAI) which is based on CSRs/WSRs and connects to the masking accelerator (MA).
This MA implements a collection of core functions in hardware which are particularly useful when implementing masking countermeasures for hardening OTBN applications against SCA, and which themselves are challenging to harden against SCA.
Internally, the MA uses a heavily pipelined first-order masking countermeasure combined with shuffling to deter SCA.
The masking countermeasure is formally verified under the transition- and glitch-extended probing model.
For details, refer to our [CocoAlma](https://github.com/lowRISC/opentitan/blob/master/hw/ip/otbn/pre_sca/alma/README.md#formally-verifying-the-mask-accelerator-interface-modules) and [PROLEAD](https://github.com/lowRISC/opentitan/tree/master/hw/ip/otbn/pre_sca/prolead#otbn-masking-leakage-analysis-using-prolead) verification setups.
```

Author: Thank you for the suggestion. This makes sense.

The OTBN SW can control this accelerator through the following CSRs / WSRs (see [here](../README.md#Control-and-Status-Registers-(CSRs)) for more details):
- `MAI_CTRL`
- `MAI_STATUS`
- `MAI_RES_S0`
- `MAI_RES_S1`
- `MAI_IN0_S0`
- `MAI_IN0_S1`
- `MAI_IN1_S0`
- `MAI_IN1_S1`

The `MAI_CTRL` and `MAI_STATUS` CSRs allow OTBN software to configure and start operations.

> All other WSRs are used either to offload or retrieve the data to/from the accelerator.

Reviewer: ```suggestion
All other WSRs are used either to load data into or retrieve results from the accelerator.
```

These WSRs are 256-bit wide and are interpreted by the MAI as 8 first-order shared 32-bit elements the same way as the vectorized instructions operate on WDRs.
The `_Sx` suffix identifies the share index.

The accelerator supports the following vectorized operations:
- secAdd

>   - Securely computes additions of two Boolean shared values and presents the results again as Boolean sharings.

Reviewer 2: "computes arithmetic additions of two Boolean shared operands" the key thing is that the addition is arithmetic and not boolean...

Author: Fair point, changed it.

  - This means that the two's-complement negation of `MOD` must be absorbed into the XOR-sums.
  - `MAI_RES_S0`/`MAI_RES_S1` will hold Boolean sharings of `(inp1 + inp2) mod MOD`.
  - Processing the full vector takes 22 cycles.
    The first result is written to the output WSRs after 15 cycles.
- A2B
  - Convert values represented with an Arithmetic sharing to a Boolean shared representation.
  - `MAI_IN0_S0`/`MAI_IN0_S1` carry Arithmetic sharings `(a0, a1)` of `a`, where `a = a0 + a1 mod MOD`.
  - `MAI_IN1_S0`/`MAI_IN1_S1` are unused.
  - `MAI_RES_S0`/`MAI_RES_S1` will hold Boolean sharings `(b0, b1)` of `a`, where `b0 XOR b1 = a`.
  - Processing the full vector takes 22 cycles.
    The first result is written to the output WSRs after 15 cycles.
- B2A
  - Convert values represented with a Boolean sharing to an Arithmetic shared representation.
  - `MAI_IN0_S0`/`MAI_IN0_S1` carry Boolean sharings `(a0, a1)` of `a`, where `a = a0 XOR a1`.
  - `MAI_IN1_S0`/`MAI_IN1_S1` are unused.
  - `MAI_RES_S0`/`MAI_RES_S1` will hold arithmetic sharings `(b0, b1)` of `a`, where `b0 + b1 mod MOD = a`.
  - Internally, there is a rejection-sampling step for each element.

>     This sampling makes the execution nondeterministic.

Reviewer: Is there an upper limit on how long this can take?

Author: This depends on the rejection probability. So it could take forever. @h-filali can maybe give an probability number?

Reviewer: It is fine to not provide a number in the spec.

If it takes forever, we could abort the MAI operation by triggering a SW error?

Author: Yes. Then the secure wipe will disable the rejection sampling and it will finish in known time.

Reviewer 3: Sounds good. The probability of a single sample being rejected is max 50%. For ML-DSA this is much much lower.

Reviewer 3: I would replace execution with something like runtime to be more specific about what exactly is non-deterministic.

    If no samples must be rejected, processing the full vector would take 23 cycles.
    The first result would be written to the output WSRs after 16 cycles.

For all operations except the secAdd, the MOD WSR must stay constant as long as an execution is ongoing.
Otherwise the results are invalid.

#### MAI usage

OTBN SW can perform a masking operation by following these steps (see also illustration below):
- Poll until `MAI_STATUS.BUSY = 0`.
- Write the input data to the input WSRs `MAI_INx_Sy`(where `x` represents the input, `y` the share index).
- Write the desired operation type to `MAI_CTRL.OPERATION`.
  - This step can also come before the data is transferred.
- Issue the `MAI_CTRL.START` command by writing a 1 to it.
  - The MAI will then start to push all 8 elements of the input WSRs sequentially into the masking accelerator.
  - As soon as the first result is available, the MAI will begin to overwrite the `MAI_RES_S0` / `MAI_RES_S1` WSRs with the new data.

> - Poll until `MAI_STATUS.BUSY = 0`.

Reviewer 2: Question: we don't have a VALID bit right now. Is there any occasion when BUSY goes low or remains low without the result being valid?

Author: No. The MAI will always produce a valid result. Any error results in a OTBN SW error.

Reviewer 2: Okay great.

- Read back the results from `MAI_RES_S0` / `MAI_RES_S1`.
- Repeat this sequence for the next data.

This sequence can be optimized, because:
- The input WSRs can be overwritten with the next data as soon as the MAI pushed all elements into the accelerator.
  - This is the case as soon as `MAI_STATUS.READY = 1`.
  - Note that `MAI_STATUS.READY = 1` does not signal that the MAI is ready to start the next execution.
    It just indicates that it is ready to receive new data.
- The START command of the next execution can be issued before the results are read back.
  - This is possible because the accelerator latency delays the update to the result WSRs.

>     See operation details for an exact latency numbers.

Reviewer: ```suggestion
    See operation details for exact latency numbers.
```

  - The START command can be issued as soon as `MAI_STATUS.BUSY = 0`.

> This allows to overlap the data transfer with the accelerator latency.

Reviewer: ```suggestion
This allows overlapping data transfer with the accelerator latency.
```

The overlapping execution software flow is:
- Setup the MAI as described above.
- Start the first execution by issuing the `MAI_CTRL.START` command.
- Poll until `MAI_STATUS.READY = 1`.
- Update the `MAI_INx_Sy` with the next data.
- Poll until `MAI_STATUS.BUSY = 0`.
- Issue the `MAI_CTRL.START` command for the next execution.

> - Read back the result of the first execution.

Reviewer 2: Do we have a bit indicating when software isn't fast enough to read the result, i.e., if the result got overwritten? I think this would add lots of verification complexity.

Author: No there is no such flag. I think this responsibility should be delegated to the SW. The minimal latency when this case happens is deterministic.

Reviewer 2: I agree. It simplifies the design a lot.

- Repeat the steps starting with "Poll until `MAI_STATUS.READY = 1`".

This procedure can be even further optimized for secAdd, secAddMod, and A2B because their latency is deterministic and thus the polling can be replaced with the correct number of other instructions.

> The following image depict the flow for multiple secAdd executions without any memory operations (to load data from/to WDRs / DMEM).

Reviewer: ```suggestion
The following image depicts the flow for multiple secAdd executions without any memory operations (to load data from/to WDRs / DMEM).
```

A result with “1p” means that this WSR contains partial results (vector elements) / is being overwritten with results from the current conversion (2p for results of the second conversion, etc).
The dummy instruction between the two `bn.wsrr` instructions is to avoid leakage when reading the two shares of the results.

In the example, the latency of the secAdd accelerator is assumed to be 5 cycles.
The unused cycles can be used by OTBN to do other tasks like loading the next inputs.

![OTBN MAI execution](./mai_execution.svg)

#### MAI error handling
An erroneous use of the MAI will result in a `MAI_SOFTWARE_ERROR`.
This will trigger an abortion of the OTBN execution with a secure wipe.

The `MAI_SOFTWARE_ERROR` is triggered when:
- `MAI_CTRL.OPERATION` contains an invalid selection when the `MAI_CTRL.START` bit is set.
- The `MAI_CTRL.START` bit is set whilst `MAI_STATUS.BUSY = 1`.
- A write to input WSRs is detected whilst `MAI_STATUS.READY = 0`.
  - The `MAI_INx_Sy` registers are used as input buffers.
    Modifying the register content can lead to unexpected results.

> - A write to the reserved section of `MAI_CTRL` is detected..

Reviewer: ```suggestion
- A write to the reserved section of `MAI_CTRL` is detected.
```

> - A write to the reserved section of `MAI_CTRL` is detected..

Reviewer 2: Remove one of the full stops.

A write to the output WSRs whilst an operation is ongoing is on purpose not handled as an error.

> This allows to clear the output WSRs between two overlapping operations if this would be required for security reasons.

Reviewer: ```suggestion
This allows clearing the output WSRs between two overlapping operations if this would be required for security reasons.
``

It is the SW's responsibility to make sure not to overwrite results.

#### MAI secure wipe behaviour

> When a secure wipe starts, any ongoing execution is let finishing.

Reviewer: ```suggestion
When a secure wipe starts, any ongoing execution is allowed to finish.
```

This MA implements a collection of core functions in hardware which are particularly useful when implementing masking countermeasures for hardening OTBN applications against SCA, and which themselves are challenging to harden against SCA.
Internally, the MA uses a heavily pipelined first-order masking countermeasure combined with shuffling to deter SCA.
The masking countermeasure is formally verified under the transition- and glitch-extended probing model.
For details, refer to our [CocoAlma](https://github.com/lowRISC/opentitan/blob/master/hw/ip/otbn/pre_sca/alma/README.md#formally-verifying-the-mask-accelerator-interface-modules) and [PROLEAD](https://github.com/lowRISC/opentitan/tree/master/hw/ip/otbn/pre_sca/prolead#otbn-masking-leakage-analysis-using-prolead) verification setups.

The OTBN SW can control this accelerator through the following CSRs / WSRs (see [here](../README.md#Control-and-Status-Registers-(CSRs)) for more details):
- `MAI_CTRL`
- `MAI_STATUS`
- `MAI_RES_S0`
- `MAI_RES_S1`
- `MAI_IN0_S0`
- `MAI_IN0_S1`
- `MAI_IN1_S0`
- `MAI_IN1_S1`

The `MAI_CTRL` and `MAI_STATUS` CSRs allow OTBN software to configure and start operations.

> All other WSRs are used either to load data into or retrieve results from the accelerator.

Reviewer 3: NIT: This suggests that `MAI_CTRL` and `MAI_STATUS` are WSRs

This sequence can be optimized, because:
- The input WSRs can be overwritten with the next data as soon as the MAI pushed all elements into the accelerator.
  - This is the case as soon as `MAI_STATUS.READY = 1`.
  - Note that `MAI_STATUS.READY = 1` does not signal that the MAI is ready to start the next execution.
    It just indicates that it is ready to receive new data.
- The START command of the next execution can be issued before the results are read back.
  - This is possible because the accelerator latency delays the update to the result WSRs.
    See operation details for exact latency numbers.
  - The START command can be issued as soon as `MAI_STATUS.BUSY = 0`.

This allows overlapping data transfers with the accelerator latency.
The overlapping execution software flow is:
- Setup the MAI as described above.
- Start the first execution by issuing the `MAI_CTRL.START` command.
- Poll until `MAI_STATUS.READY = 1`.
- Update the `MAI_INx_Sy` with the next data.
- Poll until `MAI_STATUS.BUSY = 0`.
- Issue the `MAI_CTRL.START` command for the next execution.
- Read back the result of the first execution.
- Repeat the steps starting with "Poll until `MAI_STATUS.READY = 1`".

> This procedure can be even further optimized for secAdd, secAddMod, and A2B because their latency is deterministic and thus the polling can be replaced with the correct number of other instructions.

Reviewer 3: Perfect!

A result with “1p” means that this WSR contains partial results (vector elements) / is being overwritten with results from the current conversion (2p for results of the second conversion, etc).
The dummy instruction between the two `bn.wsrr` instructions is to avoid leakage when reading the two shares of the results.

In the example, the latency of the secAdd accelerator is assumed to be 5 cycles.
The unused cycles can be used by OTBN to do other tasks like loading the next inputs.

![OTBN MAI execution](./mai_execution.svg)

#### MAI error handling
An erroneous use of the MAI will result in a `MAI_SOFTWARE_ERROR`.
This will trigger an abortion of the OTBN execution with a secure wipe.

The `MAI_SOFTWARE_ERROR` is triggered when:
- `MAI_CTRL.OPERATION` contains an invalid selection when the `MAI_CTRL.START` bit is set.
- The `MAI_CTRL.START` bit is set or `MAI_CTRL.OPERATION` is changed whilst `MAI_STATUS.BUSY = 1`.
- A write to input WSRs is detected whilst `MAI_STATUS.READY = 0`.
  - The `MAI_INx_Sy` registers are used as input buffers.
    Modifying the register content can lead to unexpected results.

> - A write to the reserved section of `MAI_CTRL` is detected.

Reviewer 3: What about the modulus?

Author: Good point. The modulus is not checked for changes. As of now SW must ensure this.

Will document this in a follow up.

Author: This is already documented in line 846 of this PR. But I clarified it in https://github.com/lowRISC/opentitan/pull/30524
